nibabel/loadsave.py

```diff
@@ -20,6 +20,54 @@
 from .deprecated import deprecate_with_version
 
 
+def _signature_matches_extension(filename, sniff):
+    """Check if signature aka magic number matches filename extension.
+
+    Parameters
+    ----------
+    filename : str or os.PathLike
+        Path to the file to check
+
+    sniff : bytes or None
+        First bytes of the file. If not `None` and long enough to contain the
+        signature, avoids having to read the start of the file.
+
+    Returns
+    -------
+    matches : bool
+       - `True` if the filename extension is not recognized (not .gz nor .bz2)
+       - `True` if the magic number was successfully read and corresponds to
+         the format indicated by the extension.
+       - `False` otherwise.
+    error_message : str
+       An error message if opening the file failed or a mismatch is detected;
+       the empty string otherwise.
+
+    """
+    signatures = {
+        ".gz": {"signature": b"\x1f\x8b", "format_name": "gzip"},
+        ".bz2": {"signature": b"\x42\x5a\x68", "format_name": "bzip2"}
+    }
+    filename = _stringify_path(filename)
+    *_, ext = splitext_addext(filename)
+    ext = ext.lower()
+    if ext not in signatures:
+        return True, ""
+    expected_signature = signatures[ext]["signature"]
+    if sniff is not None and len(sniff) >= len(expected_signature):
+        found_signature = sniff[:len(expected_signature)]
+    else:
+        try:
+            with open(filename, "rb") as fh:
+                found_signature = fh.read(len(expected_signature))
+        except OSError:
+            return False, f"Could not read file: {filename}"
+    if found_signature == expected_signature:
```

Comment:
Maybe odd styling preference, but I like to do:

```Python
if data not ready:
    prep_data
proceed as usual
```

An `else` block is a thing to keep in mind, so I tend to look for ways to avoid them. Here I'd say:

```suggestion
    if sniff is None or len(sniff) < len(expected_signature):
        try:
            with open(filename, "rb") as fh:
                sniff = fh.read(len(expected_signature))
        except OSError:
            return False, f"Could not read file: {filename}"
    if not sniff.startswith(expected_signature):
```

Reply:
sounds good!